### PR TITLE
dump memory snapshot to analyze OOMs

### PR DESCRIPTION
**Pull request by weifengpy**
when setting `enable_memory_snapshot = true` in `.toml`
* dump memory snapshots in case of OOMs. output folder is `memory_snapshot/iteration_x_exit` 
* dump regularly according to `profile_freq`. output folder is `memory_snapshot/iteration_x`
* existing `.toml` works since `enable_memory_snapshot=False` by default

snapshot is an example of the dump when OOM happens

Screenshot 2024-06-12 at 9 26 53 PM

### CODE REVIEW

File: torchtitan/profiling.py [hunk 1]
```diff
@@ -70,3 +79,62 @@ def trace_handler(prof):
...
+                )
+                torch.distributed.barrier()
+
+        logger.info(f"Memory profiler active. Snapshot will be saved at {snapshot_dir}")
+        profiler = MemoryProfiler(global_step, config.profiling.profile_freq)
+        try:
+            yield profiler
+        finally:
+            # dump snapshot when CUDA OOMs
+            profiler.step(exit_ctx=True)
```
Comment (weifengpy) [inline torchtitan/profiling.py:128]: In case of any OOM or other exceptions inside training loop, `try...finally` makes sure we still dump the memory snapshot

File: train_configs/debug_model.toml [hunk 2]
```diff
@@ -9,6 +9,7 @@ use_for_integration_test = true
 enable_profiling = true
 save_traces_folder = "profile_trace"
 profile_freq = 10
+enable_memory_snapshot = false
```
Comment (weifengpy) [inline train_configs/debug_model.toml]: existing `.toml` without `enable_memory_snapshot` still works.  `enable_memory_snapshot` is optional with `getattr(config.profiling, 'enable_memory_snapshot', False)` I am just adding it here so people can start toggle it

File: torchtitan/profiling.py [hunk 3]
```diff
@@ -15,6 +16,14 @@
 # the number of warmup steps before the active step in each profiling cycle
 WARMUP = 3
 
+# how much memory allocation/free ops to record in memory snapshots
+MEMORY_SNAPSHOT_MAX_ENTRIES = 100000
```
Comment (weifengpy) [inline torchtitan/profiling.py:20]: `MEMORY_SNAPSHOT_MAX_ENTRIES` controls how large `.pickle` can be. Right now it's `36MB`

File: torchtitan/profiling.py (hunk 1 above)
Comment (awgu) [reply torchtitan/profiling.py:128]: I wonder if we want to dump it on all exceptions or only specifically OOM exceptions.

File: torchtitan/profiling.py [hunk 4]
```diff
@@ -70,3 +79,62 @@ def trace_handler(prof):
...
+                    dir_name = f"iteration_{curr_step}_exit"
+                curr_snapshot_dir = os.path.join(snapshot_dir, dir_name)
+                if not os.path.exists(curr_snapshot_dir):
+                    os.makedirs(curr_snapshot_dir, exist_ok=True)
+                logger.info(f"Dumping memory snapshot at step {curr_step}")
+                begin = time.monotonic()
+                with open(
+                    f"{curr_snapshot_dir}/rank{rank}_memory_snapshot.pickle", "wb"
+                ) as output:
+                    pickle.dump(torch.cuda.memory._snapshot(), output)
```
Comment (fegin) [inline torchtitan/profiling.py:117]: Maybe add a threshold to control dumping the memory snapshot when the memory usage is larger than the threshold to avoid overwhelming data?

File: torchtitan/profiling.py [hunk 5]
```diff
@@ -15,6 +16,14 @@
 # the number of warmup steps before the active step in each profiling cycle
 WARMUP = 3
 
+# how much memory allocation/free ops to record in memory snapshots
+MEMORY_SNAPSHOT_MAX_ENTRIES = 100000
+
+# default memory snapshot folder
+ENABLE_MEMORY_SNAPSHOT_KEY = "enable_memory_snapshot"
+MEMORY_SNAPSHOT_FOLDER_KEY = "memory_snapshot_folder"
+MEMORY_SNAPSHOT_FOLDER_DEFAULT_VALUE = "memory_snapshot"
```
Comment (tianyu-l) [inline torchtitan/profiling.py]: We should make these into configs. Please refer to how torch_profiler does this part, e.g. put into config_manager.py

File: train_configs/debug_model.toml (hunk 2 above)
Comment (tianyu-l) [reply train_configs/debug_model.toml]: ditto: we should put default option `False` into config_manager, and remove this option in all the toml config files. Maybe only enable it to `True` in debug_model.

**Review by tianyu-l — CHANGES_REQUESTED** [review]
This is a great addition to torchtitan! Had some comments on how to structure the configs.

Also, I wonder if it makes sense to have a very short tutorial on how to read/parse the output of memory profiler. Maybe extract part of this [tutorial](https://pytorch.org/blog/understanding-gpu-memory-1/).

File: torchtitan/profiling.py (hunk 4 above)
Comment (weifengpy) [reply torchtitan/profiling.py:117]: do you mean in `MB` threashold? Right now it's bounded by number of free/allocate MEMORY_SNAPSHOT_MAX_ENTRIES. For `MB`, I can google around

File: torchtitan/profiling.py (hunk 5 above)
Comment (weifengpy) [reply torchtitan/profiling.py]: good to know config_manager.py. I will move deafults into `config_manager`

**Comment by weifengpy** [comment]
convert to draft now and will publish again after moving default into config_manager.py. But current version is good for benchmarking float8 + compile + fsdp2 on MAST

File: torchtitan/profiling.py [hunk 6]
```diff
@@ -70,3 +79,62 @@ def trace_handler(prof):
...
+                self.freq = freq
+
+            def step(self, exit_ctx: bool = False):
+                if not exit_ctx and self.step_num % self.freq != 0:
+                    self.step_num += 1
+                    return
+                if not exit_ctx:
+                    curr_step = self.step_num
+                    self.step_num += 1
+                    dir_name = f"iteration_{curr_step}"
```
Comment (tianyu-l) [inline torchtitan/profiling.py]: torch.profiler starts from step 0, whereas train.py starts from step 1. In order to make things work as expected, I suggest we do the following, so that if we set `profile_freq=10` and run training for 10 steps, there will be memory snapshots for `iteration_10` (similar to torch.profiler) and `iteration_10_exit`. I've tested this offline.
```suggestion
                self.step_num += 1
                if not exit_ctx and self.step_num % self.freq != 0:
                    return
                if not exit_ctx:
                    curr_step = self.step_num
                    dir_name = f"iteration_{curr_step}"
```

**Review by tianyu-l — COMMENTED** [review]
The overhead from `torch.profiler` is only around the steps where dumping actually happens (warmup steps + actual profiling steps).

If `_record_memory_history` is always enabled for entire training, there will constantly be overhead from this memory profiler.

In other words, `torch.profiler` only profile one step per `freq` steps, while `MemoryProfiler` profiles every step and dump all `freq` iterations per `freq` steps.

File: torchtitan/profiling.py (hunk 1 above)
Comment (weifengpy) [reply torchtitan/profiling.py:128]: good suggestion. I am catching OOM now `except torch.OutOfMemoryError` and dropped `finally`

File: torchtitan/profiling.py (hunk 4 above)
Comment (weifengpy) [reply torchtitan/profiling.py:117]: googled for `MB` threshold but did not find something useful. Currently `MEMORY_SNAPSHOT_MAX_ENTRIES=100000` conroled the file size to `36MB`. Let me know if this is still a blocker

File: torchtitan/profiling.py (hunk 6 above)
Comment (weifengpy) [reply torchtitan/profiling.py]: thanks for pointing out the difference. updated accordingly

File: train_configs/debug_model.toml [hunk 7]
```diff
@@ -9,6 +9,7 @@ use_for_integration_test = true
 enable_profiling = true
 save_traces_folder = "profile_trace"
 profile_freq = 10
+enable_memory_snapshot = true
```
Comment (tianyu-l) [inline train_configs/debug_model.toml:12]: nit: let's put the folder here as well to be consistent and informative

File: torchtitan/config_manager.py [hunk 8]
```diff
@@ -100,6 +100,18 @@ def __init__(self):
...
             help="How often to collect profiler traces, in iterations",
         )
+        self.parser.add_argument(
+            "--profiling.enable_memory_snapshot",
+            action="store_true",
+            default=False,
+            help="Whether to dump memory snapshot",
+        )
+        self.parser.add_argument(
+            "--profiling.memory_snapshot_folder",
```
Comment (tianyu-l) [inline torchtitan/config_manager.py]: nit: please rename it `save_memory_snapshot_folder` to be consistent with `save_traces_folder` and `save_tb_folder`.

File: torchtitan/config_manager.py [hunk 9]
```diff
@@ -100,6 +100,18 @@ def __init__(self):
...
+            type=str,
+            default="memory_snapshots",
```
Comment (tianyu-l) [inline torchtitan/config_manager.py]: nit: let's call it `memory_snapshot`

**Review by tianyu-l — APPROVED** [review]
great work!! thank you!
please address my nits before merge :)

**Comment by weifengpy** [comment]
> great work!! thank you! please address my nits before merge :)

thanks. will address feedback before merging

File: train_configs/debug_model.toml (hunk 7 above)
Comment (weifengpy) [reply train_configs/debug_model.toml:12]: added `save_memory_snapshot_folder` in .toml
